sqlitecluster/SQLiteNode.cpp

```diff
@@ -2580,6 +2541,77 @@ void SQLiteNode::_recvSynchronize(Peer* peer, const SData& message) {
         throw "commits remaining at end";
 }
 
+bool SQLiteNode::peerClosenessSort(const Peer* lhs, const Peer* rhs)
+{
+    // Null pointers are never fast/close.
+    if (!lhs) {
+        return false;
+    }
+    if (!rhs) {
+        return true;
+    }
+
+    // If two peers have the same latency, then the one with the higher CommitCount is closest. This case is most
+    // likely in the event that we have no latency data for either peer.
+    if (lhs->latency == rhs->latency) {
+        return SToInt((*lhs)["CommitCount"]) > SToInt((*rhs)["CommitCount"]);
+    }
+
+    // If only one peer has 0 latency, the other is faster/closer.
+    if (!lhs->latency) {
+        return false;
+    }
+    if (!rhs->latency) {
+        return true;
+    }
+
+    // Both peers have latency data, and neither are 0.
+    return lhs->latency < rhs->latency;
+}
+
+SQLiteNode::Peer* SQLiteNode::updateSyncPeer()
+{
+    // Sets are stored in sorted order. This set uses peerClosenessSort as its comparator function.
+    set<Peer*, decltype(&peerClosenessSort)> latencies {peerClosenessSort};
+
+    uint64_t commitCount = _db.getCommitCount();
+    for (auto peer : peerList) {
+        // If any of these conditions are true, then we can't use this peer.
+        if (peer->test("Permaslave") ||                 // Peer is a permaslave
+            !peer->test("LoggedIn") ||                  // Peer is not logged in.
+            !SIEquals((*peer)["State"], "SLAVING") ||   // Peer is not slaving.
+            peer->calcU64("CommitCount") <= commitCount // Peer does not have a greater commit count than us.
+            ) {
+            continue;
+        }
+
+        // We insert in sorted order (see peerClosenessSort for details).
+        latencies.insert(peer);
+    }
+    
+    // The front one is the lowest.
+    Peer* newSyncPeer = latencies.empty() ? 0 : *latencies.begin();
+
+    // Log that we've changed peers.
+    if (_syncPeer != newSyncPeer) {
+        string from, to;
+        for (Peer* p : {_syncPeer, newSyncPeer}) {
+            string* dest = (p == _syncPeer) ? &from : &to;
+            if (p) {
+                *dest = p->name + " (CommitCount=" + (*p)["CommitCount"] + ", latency=" + to_string(p->latency) + "us)";
+            } else {
+                *dest = "(NONE)";
+            }
+        }
+        SINFO("Updating SYNCHRONIZING peer from " << from << " to " << to << ".");
+
+        // And save the new sync peer internally.
+        _syncPeer = newSyncPeer;
+    }
+
+    return newSyncPeer;
```

Comment:
This could also be `return _syncPeer;` right?  This is a sort of weird function, to possibly update an attribute, and then return it.  Like, anybody who can call this function, can also just access that attribute.  It makes me wonder if we can just get rid of `_syncPeer` entirely.